Add standalone NVENC encoder

ns6089 opened the pull request:
## Description
Add standalone NVENC encoder for reference frames invalidation right now. And possibly for VFR-like bitrate adjustment and more somewhere down the line. Windows version is fully functional . Linux cuda (and possibly opengl) support is out of scope for this PR, but can easily be done later (at least the encoder side),

### In later PRs
- Investigate and test the viability of "increased vbv" option. Ideally, P-frames should not steal bitrate budget from future frames, maybe this can be achieved with vbv offset in encoder.
- Investigate the viability of using multiple ref frames (L0 > 1) since we apply strict limits on the vbv. Maybe nvenc can intelligently pick blocks from previous frames if they have higher qp? Can't imagine this being free though.
- Run VMAF benchmark for const-qp mode (document the process in .md file), and pick default values for min-qp
- Check GFE default ref frame buffer values, particularly for h264 (level4 vs level5)
- New configuration page and documentation

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I want maintainers to keep my branch updated

---
### Things to implement before merge (may be expanded)
- [x] Refactor colospace selection logic, it has too much duplication right now
- [x] Decide how to handle encoder ref frames buffer size, currently it's at 16 frames (dynamically based on resolution and framerate? through configuration?)
Result: use 5 ref frames **buffer**, which corresponds to DPB=5 in h264 terms, and DPB=6 in HEVC. Both are required minimums that must be supported by decoder profiles for given resolution.
Update: h265 level 4 only supports 4 ref frames, but in this case the client can set the limit, this level is very outdated nowadays
- [x] Investigate if NVENC can be told explicitly which ref frame to use, this can allow wider decoder support for ref frames invalidation.
Verdict: when L0 (forward prediction) list size is 1, it should always use last frame as reference,
- [x] Try to patch num_ref_frames in SPS header
Verdict: may be possible for h264, close to impossible for HEVC, either way too much hassle
- [x] Test encoder caps
NV_ENC_CAPS_SUPPORT_CABAC
NV_ENC_CAPS_WIDTH_MAX
NV_ENC_CAPS_HEIGHT_MAX
NV_ENC_CAPS_SUPPORT_CUSTOM_VBV_BUF_SIZE
NV_ENC_CAPS_SUPPORT_REF_PIC_INVALIDATION
NV_ENC_CAPS_SUPPORT_YUV444_ENCODE
NV_ENC_CAPS_SUPPORT_10BIT_ENCODE
NV_ENC_CAPS_SUPPORT_MULTIPLE_REF_FRAMES
NV_ENC_CAPS_SUPPORT_QPELMV
- [x] Switch to nv-codec-headers
- [x] Update Linux and MacOS platforms to new structures
- [x] ~~Use Peak-Signal-to-Noise-Ratio (Y-PSNR), Structural Similarity Index (Y-SSIM), and Video Multimethod Assessment Fusion (VMAF) for default min qp values~~ Not in this PR
- [x] ~~Add configuration page and documentation~~ Not in this PR

### From review
- [x] Look into entropyCodingMode encoding parameter
- [x] Proper cleanup in create_encoder()
- [x] Look into last_encoder_probe_supported_invalidate_ref_frames, if needs to be changed in multiple places
- [x] Lock h264 into High profile

cgutman commented on src/nvenc/nvenc_base.cpp:
```diff
@@ -0,0 +1,323 @@
+      return false;
+    }
+
+    BOOST_LOG(info) << "Created NvENC encoder at " << quality_preset_string_from_guid(init_params.presetGUID) << (config.vbv_multiplier > 1.0f ? " with modified VDV" : "");
```
typo: `VDV` -> `VBV`

cgutman commented on src/platform/windows/display_vram.cpp:
```diff
@@ -830,6 +755,202 @@ namespace platf::dxgi {
+      auto nvenc_colorspace = nvenc::nvenc_colorspace_from_rtsp_csc(client_config.encoderCscMode, client_config.dynamicRange && display->is_hdr());
+
+      nvenc::nvenc_config nvenc_config;
+      if (config::video.nv.nv_preset) nvenc_config.quality_preset = *config::video.nv.nv_preset - 11;
```
What's the `- 11` for?

cgutman commented on src/platform/windows/display_vram.cpp:
```diff
@@ -830,6 +755,202 @@ namespace platf::dxgi {
+
+    int
+    convert(platf::img_t &img_base) override {
+      return base.convert(img_base);
```
Nice refactoring. This will make it pretty simple if we want to have NVENC directly consume RGB frames or if we need to go that route with AMF for AMD cards.

cgutman commented on src/nvenc/nvenc_base.cpp:
```diff
@@ -0,0 +1,323 @@
+        else {
+          format_config.maxNumRefFrames = 5;
+        }
+        format_config.numRefL0 = NV_ENC_NUM_REF_FRAMES_1;
```
We should also populate `entropyCodingMode` here

cgutman commented on src/nvenc/nvenc_base.cpp:
```diff
@@ -0,0 +1,323 @@
+          format_config.maxNumRefFrames = client_config.numRefFrames;
+        }
+        else {
+          format_config.maxNumRefFrames = 5;
```
It looks like 0 is default. What does NVENC pick if you pass 0 here?

cgutman commented on src/nvenc/nvenc_base.cpp:
```diff
@@ -0,0 +1,323 @@
+    init_params.maxEncodeWidth = max_width;
+    init_params.maxEncodeHeight = max_height;
+    init_params.presetGUID = quality_preset_guid_from_number(config.quality_preset);
+    init_params.tuningInfo = NV_ENC_TUNING_INFO_ULTRA_LOW_LATENCY;
```
Can you plumb in this `tuningInfo` value from the config? We expose this option to users as `config::video.nv.nv_tune`.

ns6089 replied:
Close to final concept image with the new options page https://github.com/LizardByte/Sunshine/pull/1427#issuecomment-1640116140

cgutman commented on src/nvenc/nvenc_base.cpp:
```diff
@@ -0,0 +1,323 @@
+    }
+
+    NV_ENC_CONFIG enc_config = preset_config.presetCfg;
+    enc_config.profileGUID = NV_ENC_CODEC_PROFILE_AUTOSELECT_GUID;  // TODO: select specific profile?
```
Yeah, we should select specific profiles. The specific values should be:
- `videoFormat == 0` -> `NV_ENC_H264_PROFILE_HIGH_GUID `
- `videoFormat == 1 && dynamicRange == 0` -> `NV_ENC_HEVC_PROFILE_MAIN_GUID`
- `videoFormat == 1 && dynamicRange == 1` -> `NV_ENC_HEVC_PROFILE_MAIN10_GUID`

cgutman replied:
H.264 Main or Baseline should work, but I think we should stick to H.264 High since that's what all the other encoders do. We manually specify the profile for the FFmpeg-based encoders to ensure we always get what we expect.

ns6089 replied:
Forced H264 to High, the rest should be fine on auto.

cgutman commented on src/nvenc/nvenc_base.cpp:
```diff
@@ -0,0 +1,323 @@
+
+  nvenc_encoded_frame
+  nvenc_base::encode_frame(uint64_t frame_index, bool force_idr) {
+    if (!encoder || !output_bitstream) {
```
Can this actually happen or are you just being cautious?

ns6089 replied:
Done.

cgutman commented on src/video.cpp:
```diff
@@ -2003,6 +2172,8 @@ namespace video {
             active_hevc_mode = 0;
           }
 
+          last_encoder_probe_supported_invalidate_ref_frames = encoder->details->supports_invalidate_ref_frames;
```
We need to set this in the other paths that assign `chosen_encoder` too.

ns6089 replied:
Will look into it.

ns6089 commented:
@cgutman, need your opinion on colorspace refactoring https://github.com/LizardByte/Sunshine/pull/1427/commits/a2f34ab3de3d8f18e13ae9960d76217b52ec0d8a. When you have time, it's not blocking me.

ns6089 commented:
> @cgutman, need your opinion on colorspace refactoring [a2f34ab](https://github.com/LizardByte/Sunshine/commit/a2f34ab3de3d8f18e13ae9960d76217b52ec0d8a). When you have time, it's not blocking me.

Or better wait until I finish the Linux part, it approaches things slightly differently and might need adjustments.

ns6089 commented:
New options page (not yet pushed to this PR). Some things might still be slightly reworded, and need to run a few tests for better default QP values. But conceptually it should be done, and I think I will be able to backport it to ffmpeg nvenc backend so we don't end up with two pages.
![2023-07-18 15_21_35-Sunshine — Mozilla Firefox](https://github.com/LizardByte/Sunshine/assets/61738816/560d8ec3-a6c7-4f37-bbde-a7367cc4b77b)

cgutman commented on src/platform/linux/vaapi.cpp:
```diff
@@ -398,8 +398,8 @@ namespace va {
     void
-    set_colorspace(std::uint32_t colorspace, std::uint32_t color_range) override {
-      sws.set_colorspace(colorspace, color_range);
+    apply_colorspace() override {
```
This function takes a parameter, right?

cgutman replied:
Nevermind

cgutman commented:
LGTM. Let's try to get this refactoring in soon if possible. I've got some AV1 and client-side cursor changes that I'd like to work on, but I don't want to stomp all over your work.

If you can back out any debugging changes or WIP test changes (the scheduling priority revert?), I can test this across all my systems and we can get it merged.

ns6089 commented:
> LGTM. Let's try to get this refactoring in soon if possible. I've got some AV1 and client-side cursor changes that I'd like to work on, but I don't want to stomp all over your work.
> 
> If you can back out any debugging changes or WIP test changes (the scheduling priority revert?), I can test this across all my systems and we can get it merged.

Understandable. Wired the encoder to use the existing configuration page, since some new options pretty much depend on whether I can make it work with realtime priority (and priority selection itself will add another option).

Haven't tested native HDR (only SDR in BT.2020) and CUDA path on Linux (didn't make drastic changes, but there's always a chance). Other than that, should be alright to merge in its current state. The rest of the features shouldn't produce conflicts, and can be done in later PRs.

ReenigneArcher commented on third-party/nv-codec-headers:
This [commit](https://github.com/FFmpeg/nv-codec-headers/commit/855f8263d97bbdcaeabaaaa2997e1ccad7c52dc3) is not on the branch we are tracking -> https://github.com/FFmpeg/nv-codec-headers/tree/sdk/11.1.

Can we change the branch to sdk/12.0?

ns6089 replied:
I think simply doing this https://github.com/LizardByte/Sunshine/pull/1427/commits/eafa29788da80a2961833fc6d9edeb7ee4549beb is enough

ns6089 replied:
> changing the branch in the got modules file is not enough. You also must checkout the latest commit on that branch and commit it.

Alright. Rebased all of that into one commit https://github.com/LizardByte/Sunshine/pull/1427/commits/7a336ca3593a72a82535f7ff49fed4de7e7c964c

ns6089 replied:
And it now fails the build...

ns6089 replied:
Turns out I used one status constant from 12.1 in pretty debug print function. Should be fine now.

cgutman commented on src/nvenc/nvenc_base.cpp:
```diff
@@ -0,0 +1,541 @@
+
+    if (last_frame < first_frame ||
+        encoder_state.last_encoded_frame_index < first_frame ||
+        encoder_state.last_encoded_frame_index > first_frame + 100) {
```
That last condition (`encoder_state.last_encoded_frame_index > first_frame + 100`) is actually possible if we have a period of severe packet loss. We will continue expanding our RFI range until we successfully receive a recovery frame from the host. We shouldn't log an error for that case.

ns6089 replied:
The reason why it's there is because nvenc rfi api doesn't work with ranges, but individual frames. And these calls can get expensive really fast if we don't limit them to some reasonable number. I will add this to my todo, some more intelligent logic can be written based on previously requested range, but it shouldn't be a blocker since it's just a log message that happens in very specific circumstances.

cgutman replied:
That's fine with me. I just mentioned it because I saw the logs when I was performing RFI and packet loss tests with this branch.

cgutman commented on src/nvenc/nvenc_utils.cpp:
```diff
@@ -0,0 +1,74 @@
+        colorspace.tranfer_function = NV_ENC_VUI_TRANSFER_CHARACTERISTIC_BT2020_10;
+        colorspace.matrix = NV_ENC_VUI_MATRIX_COEFFS_BT2020_NCL;
+        break;
+    }
```
These cases look mixed up.

ns6089 replied:
Nice catch, this is why huge commits are evil :(

cgutman replied:
I think the BT.2020 SDR and HDR cases are still reversed.

ns6089 replied:
Yep. and I don't have an excuse this time :(

cgutman commented on src/video_colorspace.cpp:
```diff
@@ -0,0 +1,174 @@
+      }
+      else {
+        // Rec. 2020
+        colorspace.colorspace = colorspace_e::bt2020sdr;
```
Moonlight expects the host to encode using its requested colorspace even when streaming in 10-bit modes, so the above case should be:
```
      if (hdr_requested && hdr_display) {
        // Rec. 2020 with ST 2084 perceptual quantizer
        colorspace.colorspace = colorspace_e::bt2020;
      }
      else {
      ...

cgutman replied:
Rec. 601 is probably correct. We generally request that as a default for historical reasons, except for some GPUs that only behave properly with Rec. 709. I'll take a look at the crash.

Btw, GFE defaults to Rec. 601 if the colorspace isn't recognized. It's just academic because Moonlight will not pass an invalid color space, but just mentioning it since you added a codepath for that.

ns6089 replied:
There was some misunderstanding on how 10-bit encoding is activated.
Pushed a fix https://github.com/LizardByte/Sunshine/pull/1427/commits/93d4b27dfdb934bc6114a163c873a27cda16456f

cgutman commented:
The latest changes look good. Once you squash the fixup commits and rebase, I'll do a final testing pass on my local machines and we can get this in.

ns6089 commented:
Squashed and rebased. Also added small "fix" for nvapi, it doesn't affect anything since we're strictly x64, so no point in opening full pull request for it.

cgutman commented:
Everything looks good in my tests:
- HDR with and without NVENC
- NvFBC on Linux
- All valid colorspace and color range combos
